File: src/Butil/Bit.Butil/Scripts/launchQueue.ts
var BitButil = (window as any).BitButil = (window as any).BitButil || {};

(function (butil: any) {
    // The launches the app was started with. `setConsumer` may only be called once per page, and the
    // browser delivers a launch as soon as a consumer exists - which can be before .NET has asked
    // for one. So the consumer is installed while this module is evaluated and the launches are
    // parked here, in arrival order, until a .NET handler shows up. A queue rather than a single
    // slot: a launch handler routing a second "open with" into this window would otherwise drop the
    // first one.
    let _pending: any[] = [];
    let _consumer: { dotNetRef: any, listenerId: string, method: string } | null = null;

    // The launched files' handles, kept so their contents can be read (and written back) later.
    // Keyed by an opaque launch id, and within a launch by the position .NET sees in the reported
    // file list - so a second launch cannot become a way to reach the first launch's files.
    const _launches: { [launchId: string]: any[] } = {};
    let _launchCount = 0;
    let _lastLaunchId = '';

    // A page that is launched over and over would otherwise hold every file handle it was ever
    // given; the recent ones are what a consumer can still be reading.
    const MAX_RETAINED_LAUNCHES = 8;
    const _launchIds: string[] = [];

    function handlesOf(launchId: string) {
        // An empty id is the by-index API talking about the most recent launch.
        return _launches[launchId || _lastLaunchId] ?? [];
    }

    const queue = (window as any).launchQueue;
    if (queue?.setConsumer) {
        try {
            queue.setConsumer((params: any) => deliver(params));
        } catch {
            // Another consumer was already set - nothing to do.
        }
    }

    async function describe(params: any) {
        const launchId = `launch-${++_launchCount}`;
        const handles = Array.isArray(params?.files) ? [...params.files] : [];

        _launches[launchId] = handles;
        _lastLaunchId = launchId;
        _launchIds.push(launchId);
        while (_launchIds.length > MAX_RETAINED_LAUNCHES) delete _launches[_launchIds.shift() as string];

        const files: any[] = [];
        for (let i = 0; i < handles.length; i++) {
            const handle = handles[i];
            let name = handle?.name ?? '';
            let size = 0;
            let type = '';
            let lastModified = 0;
            try {
                const file = await handle.getFile();
                name = file.name ?? name;
                size = file.size ?? 0;
                type = file.type ?? '';
                lastModified = file.lastModified ?? 0;
            } catch {
                // The handle went stale, or permission was revoked between launch and read.
            }
            files.push({ launchId, index: i, name, type, size, lastModified });
        }

        return { launchId, targetUrl: params?.targetURL ?? '', files };
    }

    async function deliver(params: any) {
        const payload = await describe(params);
        if (_consumer) butil.utils.dispatch(_consumer.dotNetRef, _consumer.method, _consumer.listenerId, payload);
        else _pending.push(payload);
    }

    butil.launchQueue = {
        isSupported() { return 'launchQueue' in window; },
        // File handling is a separate capability from launch handling: a runtime can deliver a
        // launch without ever putting files on it.
        supportsFiles() { return 'LaunchParams' in window && 'files' in ((window as any).LaunchParams.prototype ?? {}); },
        setConsumer(dotNetRef: any, listenerId: string, method: string) {
            _consumer = { dotNetRef, listenerId, method };
            // Launches that arrived before .NET was ready are replayed rather than dropped, oldest
            // first, so the handler sees them in the order the browser delivered them.
            const pending = _pending;
            _pending = [];
            for (const payload of pending) butil.utils.dispatch(dotNetRef, method, listenerId, payload);
        },
        clearConsumer(listenerId: string) {
            if (_consumer?.listenerId === listenerId) _consumer = null;
        },
        async readText(launchId: string, index: number) {
            const handle = handlesOf(launchId)[index];
            if (!handle?.getFile) return null;
            try {
                const file = await handle.getFile();
                return await file.text();
            } catch {
                return null;
            }
        },
        async readBytes(launchId: string, index: number) {
            const handle = handlesOf(launchId)[index];
            if (!handle?.getFile) return null;
            try {
                const file = await handle.getFile();
                return new Uint8Array(await file.arrayBuffer());
            } catch {
                return null;
            }
        },
        async writeText(launchId: string, index: number, contents: string) {
            const handle = handlesOf(launchId)[index];
            if (!handle?.createWritable) return false;
            try {
                // A file-handling launch grants read access; writing may still need an explicit
                // permission request, which createWritable triggers.
                const writable = await handle.createWritable();
                await writable.write(contents);
                await writable.close();
                return true;
            } catch {
                return false;
            }
        },
        async writeBytes(launchId: string, index: number, contents: Uint8Array) {
            const handle = handlesOf(launchId)[index];
            if (!handle?.createWritable) return false;
            try {
                const writable = await handle.createWritable();
                await writable.write(butil.utils.arrayToBuffer(contents));
                await writable.close();
                return true;
            } catch {
                return false;
            }
        }
    };
}(BitButil));

File: src/Butil/Bit.Butil/Publics/LaunchQueue.cs
using System;
using System.Collections.Concurrent;
using System.Diagnostics.CodeAnalysis;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.JSInterop;

namespace Bit.Butil;

/// <summary>
/// Wraps <see href="https://developer.mozilla.org/en-US/docs/Web/API/Window/launchQueue">window.launchQueue</see>
/// - how an installed app receives the files, and the target URL, it was launched with.
/// </summary>
/// <remarks>
/// This only ever fires for an <b>installed</b> app whose manifest declares what it handles:
/// <c>file_handlers</c> for "open with", <c>protocol_handlers</c> for a custom scheme,
/// <c>share_target</c> for a share, and <c>launch_handler</c> for how a second launch is routed
/// (a new window, or the existing one). A page open in a browser tab never receives a launch.
/// <br/>
/// <b>Timing:</b> the browser delivers the launch the moment a consumer exists, and a page may set
/// one only once. Butil installs its consumer while the <c>launchQueue</c> module is evaluated and
/// parks the launch until <see cref="SetConsumer(Action{LaunchParams})"/> is called, so a handler
/// registered in <c>OnAfterRenderAsync</c> still receives it. Under
/// <see cref="BitButil.UseLazyScripts"/> the module isn't imported until the first call into it -
/// call <see cref="IsSupported"/> early in app start-up so the consumer is in place.
/// <br/>
/// A launched file's contents stay on the JS side and are read by index - see <see cref="LaunchFile"/>.
/// </remarks>
[ButilService(typeof(LaunchQueue))]
public class LaunchQueue(IJSRuntime js) : IAsyncDisposable
{
    internal const string InvokeMethodName = nameof(InvokeLaunch);

    private readonly ConcurrentDictionary<Guid, Action<LaunchParams>> _handlers = new();

    // Per-instance callback reference (see Keyboard): the consumer is isolated per circuit / WASM
    // app and released on disposal - no static state, no cross-circuit leak.
    private DotNetObjectReference<LaunchQueue>? _dotNetRef;
    private DotNetObjectReference<LaunchQueue> DotNetRef => DotNetObjectReferenceHelper.GetOrCreate(ref _dotNetRef, this);

    /// <summary>True when the runtime exposes <c>window.launchQueue</c> (Chromium only).</summary>
    /// <remarks>
    /// During prerender/SSR (no JS runtime) this returns <c>default</c> (e.g. <c>false</c>/<c>0</c>)
    /// rather than throwing, so the result can't be distinguished from a genuine value. If you
    /// branch on it, defer the read to <c>OnAfterRenderAsync</c>.
    /// </remarks>
    public ValueTask<bool> IsSupported() => js.Invoke<bool>("BitButil.launchQueue.isSupported");

    /// <summary>
    /// True when the runtime also implements File Handling, i.e. a launch can carry files
    /// (<c>LaunchParams.files</c>). Launch handling without file handling is possible.
    /// </summary>
    public ValueTask<bool> SupportsFiles() => js.Invoke<bool>("BitButil.launchQueue.supportsFiles");

    /// <summary>
    /// Invoked from JS when the app is launched. Public + <see cref="JSInvokableAttribute"/> so it
    /// can be dispatched through the per-instance <see cref="DotNetObjectReference{T}"/>.
    /// </summary>
    [JSInvokable(InvokeMethodName)]
    public void InvokeLaunch(Guid id, LaunchParams launchParams)
    {
        if (_handlers.TryGetValue(id, out var handler)) handler.Invoke(launchParams ?? new LaunchParams());
    }

    /// <summary>
    /// Registers the handler that receives this app's launch. Fires immediately when the launch has
    /// already arrived, so registering late is safe.
    /// </summary>
    /// <param name="handler">Called once per launch, with the target URL and any files.</param>
    /// <returns>
    /// A subscription that unregisters the handler on dispose. Only the most recent registration
    /// receives launches; the underlying <c>setConsumer</c> is set once by Butil and never replaced.
    /// </returns>
    [DynamicDependency(nameof(InvokeLaunch), typeof(LaunchQueue))]
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(LaunchParams))]
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(LaunchFile))]
    public async Task<ButilSubscription> SetConsumer(Action<LaunchParams> handler)
    {
        ArgumentNullException.ThrowIfNull(handler);

        var id = Guid.NewGuid();
        _handlers.TryAdd(id, handler);

        await js.InvokeVoid("BitButil.launchQueue.setConsumer", DotNetRef, id, InvokeMethodName);

        return new ButilSubscription(id, async () =>
        {
            _handlers.TryRemove(id, out _);
            await js.InvokeVoid("BitButil.launchQueue.clearConsumer", id);
        });
    }

    /// <summary>
    /// Reads a launched file as text.
    /// </summary>
    /// <param name="file">A file from the launch.</param>
    /// <returns>The contents, or null when the handle went stale or access was refused.</returns>
    public ValueTask<string?> ReadText(LaunchFile file)
    {
        ArgumentNullException.ThrowIfNull(file);
        return ReadText(file.Index);
    }

    /// <summary>Reads a launched file as text, by its <see cref="LaunchFile.Index"/>.</summary>
    public ValueTask<string?> ReadText(int index) => js.Invoke<string?>("BitButil.launchQueue.readText", index);

    /// <summary>
    /// Reads a launched file as bytes.
    /// </summary>
    /// <returns>The contents, or null when the handle went stale or access was refused.</returns>
    public ValueTask<byte[]?> ReadBytes(LaunchFile file)
    {
        ArgumentNullException.ThrowIfNull(file);
        return ReadBytes(file.Index);
    }

    /// <summary>Reads a launched file as bytes, by its <see cref="LaunchFile.Index"/>.</summary>
    public ValueTask<byte[]?> ReadBytes(int index) => js.Invoke<byte[]?>("BitButil.launchQueue.readBytes", index);

    /// <summary>
    /// Writes text back to a launched file - the "save" half of an editor launched through
    /// <c>file_handlers</c>.
    /// </summary>
    /// <returns>
    /// False when the runtime can't write to the handle, or the user refused the write permission
    /// the browser asks for on the first save.
    /// </returns>
    public ValueTask<bool> WriteText(LaunchFile file, string contents)
    {
        ArgumentNullException.ThrowIfNull(file);
        return WriteText(file.Index, contents);
    }

    /// <summary>Writes text back to a launched file, by its <see cref="LaunchFile.Index"/>.</summary>
    public ValueTask<bool> WriteText(int index, string contents)
        => js.Invoke<bool>("BitButil.launchQueue.writeText", index, contents);

    /// <summary>Writes bytes back to a launched file.</summary>
    /// <returns>False when the runtime can't write to the handle, or the write permission was refused.</returns>
    public ValueTask<bool> WriteBytes(LaunchFile file, byte[] contents)
    {
        ArgumentNullException.ThrowIfNull(file);
        return WriteBytes(file.Index, contents);
    }

    /// <summary>Writes bytes back to a launched file, by its <see cref="LaunchFile.Index"/>.</summary>
    public ValueTask<bool> WriteBytes(int index, byte[] contents)
        => js.Invoke<bool>("BitButil.launchQueue.writeBytes", index, contents);

    /// <summary>Unregisters every handler registered through this instance and releases its interop reference.</summary>
    public async ValueTask DisposeAsync()
    {
        try
        {
            var ids = _handlers.Keys.ToArray();
            _handlers.Clear();
            foreach (var id in ids) await js.InvokeVoid("BitButil.launchQueue.clearConsumer", id);
        }
        catch (Exception ex) when (ex.IsIgnorableDisposalException()) { } // teardown: circuit gone, cancelled, or already disposed
        finally
        {
            _dotNetRef?.Dispose();
            _dotNetRef = null;
        }
        GC.SuppressFinalize(this);
    }
}
